File: src/main/java/tech/simter/reactive/io/FileUtils.java
package tech.simter.reactive.io;

import reactor.core.publisher.Flux;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.BaseStream;

/**
 * @author RJ
 */
public class FileUtils {
  /**
   * Read line by line from a file as a {@link Flux Flux&lt;String&gt;}.
   *
   * <p> This method use {@link Files#lines(Path, Charset)} indeed. It will lazily read the lines from the file,
   * without ever having to hold the whole content of the file in memory.
   *
   * <p> Bytes from the file are decoded into characters using the specified charset.
   *
   * <p> After done reading, auto close the file stream.
   *
   * @param path    the path to the file
   * @param charset the charset to use for decoding
   * @return a new {@link Flux Flux&lt;String&gt;} built around a auto close resource
   */
  public static Flux<String> readLineToString(Path path, Charset charset) {
    return Flux.using(() -> Files.lines(path, charset), Flux::fromStream, BaseStream::close);
  }

  /**
   * Read line by line from a file as a {@link Flux Flux&lt;String&gt;}.
   *
   * <p> This method works as if invoking it were equivalent to evaluating the expression:
   * <pre>{@code
   * readLineToString(path, StandardCharsets.UTF_8)
   * }</pre>
   *
   * @param path the path to the file
   * @return a new {@link Flux Flux&lt;String&gt;} built around a auto close resource
   */
  public static Flux<String> readLineToString(Path path) {
    return readLineToString(path, StandardCharsets.UTF_8);
  }
}

File: src/main/kotlin/tech/simter/reactive/io/FileUtils.kt
package tech.simter.reactive.io

import reactor.core.publisher.Flux
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

/**
 * Some reactive ways to deal with file resource.
 *
 * @author RJ
 */
object FileUtils {
  /**
   * Read line by line from a file as [Flux<String>].
   *
   * This method use [Files.lines] indeed. It will lazily read the lines from the specific [path] file,
   * without ever having to hold the whole content of the file in memory.
   * Bytes from the file are decoded into characters using the specified [charset].
   * After done reading, auto close the file stream.
   *
   * @return a new [Flux] with [String] element type built around a auto close resource
   */
  @JvmOverloads
  fun readLineToString(path: Path, charset: Charset = StandardCharsets.UTF_8): Flux<String> {
    return Flux.using({ Files.lines(path, charset) }, { Flux.fromStream(it) }, { it.close() })
  }
}